init: workarounds.nixgl module

michaelCTS opened the pull request:
### Description

The option allows packages to be wrapped by nixgl in order to make GUI applications work when installed with home-manager. The goal of this PR is just to make it easy to use graphically accelerated programs installed with home-manager in a desktop environment. 

No need to install and run `nixGL $program`. Just add the package to `workaround.nixgl.packages` like so `[ { pkg = pkgs.some_package; }]`.

Anything more advanced can come in further PRs by other people.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->


Closes #3968

michaelCTS commented on modules/misc/opengl.nix:
```diff
@@ -0,0 +1,62 @@
+{ config, pkgs, lib, ... }:
+
+let
+  nixGL = (import (builtins.fetchTarball
+    "https://github.com/nix-community/nixGL/archive/main.tar.gz") {
+      inherit pkgs;
+    });
```
This is used in the default option below, but I'm unsure how to create an option that depends on an external package and only pull the external package when a user needs it. In its current state, it always pulls it in which means `home-manager` becomes dependent on nixgl, even for users who use home-manager in a headless environment.

michaelCTS commented on modules/misc/opengl.nix:
```diff
@@ -0,0 +1,62 @@
+      default = "${nixGL.auto.nixGLDefault}/bin/nixGL";
+    };
+    packages = lib.mkOption {
+      type = lib.types.listOf lib.types.anything;
+      default = [ ];
+      example = ''[{
+        binPathRel = "bin/alacritty";
+        pkg = pkgs.alacritty;
+      }]'';
```
Dunno if a custom type is needed here of it is sufficient to use `lib.types.anything`.

michaelCTS commented:
Well, I tried adding tests, but for some reason this fails

```nix
{ lib, pkgs, ... }:                                                                           

with lib;

{
 config = {
  workarounds.nixgl.packages = with pkgs; [{ pkg = alacritty; }];                           
  
  nmt.script = ''
   assertFileRegex .nix-profile/bin/alacritty 'exec "/nix/store/.*/bin/nixGL" .*/alacritty'
   assertLinkExists .nix-profile/share/applications/Alacritty.desktop                      
  '';
 };
}
```

It doesn't find the files even though when I run `home-manage switch` I do see them.

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+    };
+in {
+  meta.maintainers = with lib.maintainers; [ michaelCTS ];
+  options.workarounds.nixgl = {
```
I assume this only applies to non-NixOS systems? In which case I think the options should go under `targets.genericLinux.nixgl`.

michaelCTS replied:
Thanks, renamed

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+        You can download the nixGL version of your choice and point to it here.
+      '';
+    };
+    wrapperBinName = lib.mkOption {
```
Please separate options by an empty line.
```suggestion

    wrapperBinName = lib.mkOption {
```

michaelCTS replied:
Applied and force-pushed.

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+    };
+    wrapperPkgPath = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      example = ''["auto" "nixGLNvidia"]'';
```
```suggestion
      example = ["auto" "nixGLNvidia"];
```

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+  options.workarounds.nixgl = {
+    src = lib.mkOption {
+      type = lib.types.package;
+      default = builtins.fetchTarball
+        "https://github.com/nix-community/nixGL/archive/main.tar.gz";
```
The user will have to provide this themselves. We cannot have the situation that HM downloads a random revision of some repository. Further, this will add a documentation dependency on this repository, i.e., pretty much all HM users would need to download this repository, whether or not they use nixGL.

```suggestion
```

michaelCTS replied:
I was hoping lazy eval wouldn't evaluate it until a user put stuff into `.packages`, but the example should give enough guidance :+1:

michaelCTS replied:
Applied and force-pushed.

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+        "Name of the nixGL binary to be called from within the wrapper package";
+    };
+    wrapperPkgPath = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
```
If the paths are stable in nixGL then perhaps this could be `enum [ "nixGLDefault" "nixGLNvidia" "nixGLNvidiaBumblebee" "nixGLIntel" ]`? The module can know the corresponding attribute path.

rycee replied:
If so then can rename the option to something like `wrapperFlavor` or something.

michaelCTS replied:
I'm not sure, honestly. There seems to also be `nixGLNvidiaBumblebee` and `nixVulkanNividia` according to the [README](https://github.com/nix-community/nixGL). If we introduced an `enum` that would make users unable to target them.

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+        You can download the nixGL version of your choice and point to it here.
+      '';
+    };
+    wrapperBinName = lib.mkOption {
```
Is the binary ever called something other than `nixGL`?

michaelCTS replied:
Yeah, seems like `nixGLNvidia`, `nixGLNvidiaBumblebee`, and others exist.

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+      makeWrapper ${wrapperBinPath} $out/bin/${bin} \
+        --add-flags ${pkg}/bin/${bin}
+    '');
+  PackageConfiguration = with lib;
```
```suggestion
  packageConfiguration = with lib;
```

michaelCTS replied:
Renamed and force-pushed

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+  };
+
+  config = {
+    home.packages =
+      builtins.map wrapWithNixGL config.workarounds.nixgl.packages;
```
Instead of wrapping each package individually, perhaps it would be less work to simply wrap the binaries in the final profile package. Not sure if it is a good idea. But I guess it would look something like this:

``` nix
home.packages = cfg.packages;
home.extraProfileCommands = ''
  for pkg in ${escapeShellArgs cfg.packages}; do
    for exe in "$out"/bin/*; do
      ${wrapWithNixGL} "$out/bin/$(basename "$exe")"
    done
  done
'';
```

rycee replied:
Ah, right you also want to wrap only a specific path in bin, then perhaps it is not worth using `extraProfileCommands`.

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+            By default this uses `package.pname`.
+          '';
+        };
+        pkg = mkOption {
```
```suggestion

        package = mkOption {
```

michaelCTS replied:
Renamed and force pushed

rycee commented on modules/misc/nixgl.nix:
```diff
@@ -0,0 +1,103 @@
+            It's expected to have a `pname` attribute.
+          '';
+        };
+      };
```
Can just set the binary option directly in the submodule:
```suggestion
      };
      
      config = {
        binary = mkOptionDefault (config.package.meta.mainProgram or config.package.pname);
      };
```

Can then remove allowing `null` values.

michaelCTS replied:
I'm not able to get this to work. It says `attribute "package" missing`. I tried `binary = mkOptionDefault (config: (config.package.meta.mainProgram or getName config.package));` hoping the function would be evaluated, but it doesn't.

rycee commented on modules/lib/maintainers.nix:
```diff
@@ -201,6 +201,12 @@
     github = "MForster";
     githubId = 4067975;
   };
+  michaelCTS = {
+    name = "michaelCTS";
+    email = "michael.vogel@cts.co";
+    github = "michaelCTS";
+    githubId = 132582212;
+  };
```
No need to add yourself here if you already is in the Nixpkgs `maintainer-list.nix`.
```suggestion
```

michaelCTS replied:
Removed and force-pushed

rycee commented:
Thanks for the contribution! I've added a bunch of, hopefully, useful comments. Note, to actually include the module you need to add it to the `modules/modules.nix` file.

exzombie commented:
I'm really happy that you found the time and will to make a pull request! Many of us have built our own hacks and workarounds, and will appreciate having a solution as part of HM.

I haven't had the time to try this yet, but just looking at the interface, I'm a bit surprised that you chose to implement a list of packages to enable nixgl for. Having followed the various discussions, I got the impression that most of us went with wrapper functions. Is there a particular reason why you chose to go with a list of packages instead?

One reason why a wrapper is more convenient is that you can wrap a package regardless of where it's used, be it in `home.packages` or in `programs.whatever.package`. As noted by @joefiorini in #3968, not being able to use nixgl in `programs.kitty.package` is a real problem.

One more question: does this module require building the home configuration with `--impure`?

michaelCTS commented:
@rycee I addressed your comments and applied the changes. 

Additionally, I added the wrapper to `config.lib.nixGl.wrap` like @Smona. That should resolves your problem @joefiorini 

```nix
programs.kitty = {                                          
 enable = true;                                            
 package = config.lib.nixGl.wrap { package = pkgs.kitty; };
};                                                          
```

> One more question: does this module require building the home configuration with `--impure`?

@exzombie I have no idea what that does. It works without that for me.

michaelCTS commented:
Closed in favor of  #5355
